// space-api/dist/lib/db/aggregate.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var db = require('./db');

var utils = require('../utils'),
    fetchAsync = utils.fetchAsync;
/**
 * Class representing the DB Aggregate Interface.
 * @example 
 * import { API, cond, or, and } from 'space-api';
 * 
 * const api = new API('my-project', 'http://localhost:4122');
 * const db = api.DB("mongo");
 * 
 * const pipe = [
 *   { $match: { status: 'A' } },
 *   { $group: { _id: '$cust_id', total: { $sum: '$amount' } } }
 * ]
 * 
 * db.aggr('posts').pipe(pipe).apply().then(res => {
 *   if (res.status === 200) {
 *     // res.data contains the documents returned by the database
 *     console.log('Response:', res.data);
 *     return
 *   }
 * }).catch(ex => {
 *   // Exception occured while processing request
 * });
 */


var Aggregate =
/*#__PURE__*/
function () {
  /**
   * Create an instance of the DB Aggregate Interface.
   * @param {string} appId 
   * @param {string} collection 
   * @param {string} url 
   * @param {Object} options 
   * @param {string} db 
   * @param {string} op 
   */
  function Aggregate(appId, collection, url, options, db, op) {
    _classCallCheck(this, Aggregate);

    this.appId = appId;
    this.collection = collection;
    this.url = url;
    this.options = Object.assign({}, options, {
      method: 'POST'
    });
    this.db = db;
    this.params = {};
    this.params.op = op;
  }
  /**
   * Prepares the Pipe query
   * @param {Object[]} pipeObj - The pipeline object.
   */


  _createClass(Aggregate, [{
    key: "pipe",
    value: function pipe(pipeObj) {
      this.params.pipe = pipeObj;
      return this;
    }
    /**
     * Makes the query to return single object. 
     * @returns {Promise} Returns a promise containing response from server.
     * @example
     * db.aggr('posts').pipe([...]).one().then(res => ...)
     * @deprecated Since version 0.4.3. Will be deleted in version 1.0.0. Use apply instead.
     */

  }, {
    key: "one",
    value: function one() {
      this.params.op = 'one';
      this.options.body = JSON.stringify(this.params);
      var url = db.dbURL(this.url, this.db, this.appId, 'crud', "".concat(this.collection, "/aggr"));
      return fetchAsync(url, this.options);
    }
    /**
     * Makes the query to return all objects. 
     * @returns {Promise} Returns a promise containing response from server.
     * @example
     * db.aggr('posts').pipe([...]).all().then(res => ...)
     * @deprecated Since version 0.4.3. Will be deleted in version 1.0.0. Use apply instead.
     */

  }, {
    key: "all",
    value: function all() {
      this.params.op = 'all';
      this.options.body = JSON.stringify(this.params);
      var url = db.dbURL(this.url, this.db, this.appId, 'crud', "".concat(this.collection, "/aggr"));
      return fetchAsync(url, this.options);
    }
    /**
     * Makes the query to return all objects. 
     * @returns {Promise} Returns a promise containing response from server.
     * @example
     * db.aggr('posts').pipe([...]).apply().then(res => ...)
     */

  }, {
    key: "apply",
    value: function apply() {
      this.options.body = JSON.stringify(this.params);
      var url = db.dbURL(this.url, this.db, this.appId, 'crud', "".concat(this.collection, "/aggr"));
      return fetchAsync(url, this.options);
    }
  }]);

  return Aggregate;
}();

module.exports = Aggregate;
